Python/Product/Analysis/LanguageServer/ParseQueue.cs

```diff
@@ -32,181 +32,223 @@ class ParseQueue {
-                                );
-                            }
-                        }
-                        if (!buffers.Any()) {
-                            throw new FileNotFoundException($"failed to parse file {pyEntry.DocumentUri.AbsoluteUri}", pyEntry.FilePath);
-                        }
-                        complete = true;
-                        result = UpdateTree(pyEntry, buffers);
-                    } finally {
-                        if (!complete) {
-                            AbortParsingTree(pyEntry);
-                        }
+        public DiagnosticSeverity InconsistentIndentation { get; set; }
+
+        sealed class ParseTask {
+            private readonly ParseQueue _queue;
+            private readonly IDocument _document;
+            private readonly PythonLanguageVersion _languageVersion;
+
+            private readonly IPythonParse _parse;
+
+            private readonly TaskCompletionSource<IAnalysisCookie> _tcs;
+            private ParseTask _next;
+
+            // State transitions:
+            //  UNSTARTED -> QUEUED     when passed to ContinueWith
+            //  UNSTARTED -> DISPOSED   when DisposeIfNotStarted() is called
+            //  UNSTARTED -> STARTED    when Start() is called
+            //  QUEUED    -> STARTED    when Start(QUEUED) is called
+            //  STARTED   -> DISPOSED   when task completes
+            // Note that calling Dispose() only has an effect on a task that
+            // has not been started or queued.
+            private const int UNSTARTED = 0;
+            private const int QUEUED = 1;
+            private const int STARTED = 2;
+            private const int DISPOSED = 3;
+            private int _state = UNSTARTED;
+
+            public ParseTask(ParseQueue queue, IDocument document, PythonLanguageVersion languageVersion) {
+                _queue = queue;
+                _document = document;
+                _languageVersion = languageVersion;
+
+                _queue._parsingInProgress.Increment();
+                _parse = (_document as IPythonProjectEntry)?.BeginParse();
+
+                _tcs = new TaskCompletionSource<IAnalysisCookie>();
+            }
+
+            public Task<IAnalysisCookie> Task => _tcs.Task;
+
+            public void DisposeIfNotStarted() {
+                if (Interlocked.CompareExchange(ref _state, DISPOSED, UNSTARTED) == UNSTARTED) {
+                    _parse?.Dispose();
+                    DisposeWorker();
+                }
+            }
+
+            private void DisposeWorker() {
+                Debug.Assert(Volatile.Read(ref _state) == DISPOSED);
+
+                var next = Interlocked.Exchange(ref _next, null);
+                _queue._parsingInProgress.Decrement();
+                _queue._parsing.TryUpdate(_document, next, this);
+                next?.Start(QUEUED);
+            }
+
+            public ParseTask ContinueWith(ParseTask nextTask) {
```

Comment:
`ContinueWith` is invoked in inside `updateValueFactory`, which means that you don't know how how many times it is called and which of results will be used.
For example, if there are two concurrent requests, and `_parsing` already has a value, `ContinueWith` may be called 3 times:
- Invocation of  `updateValueFactory` for the first request: `ContinueWith` changes `_next`  to the `ParseTask` from first request
- Invocation of  `updateValueFactory` for the second request: `ContinueWith` finds that `_next` is already set and changes `_next` of the the `ParseTask` from first request to the `ParseTask` from second request
- Second invocation processed faster than the first one, so dictionary accepts `ParseTask` from second request as a new value
- Another invocation of the `updateValueFactory` for the first request: `ContinueWith` changes `_next`  of the the `ParseTask` from second request to the `ParseTask` from first request

As a result, we have a loop.

Reply:
Ah, it's worse than that. They just use a `while (true)` loop and keep calling either factory until something succeeds. I think I can make this safer though.

Reply:
I think I've managed to fix it with smaller change: https://github.com/AlexanderSher/PTVS/commit/15fab6679bdbc759c44e32389a6f8bcf96b8ccb8